### PR TITLE
Fix llvmir and assembly

### LINKED ISSUE #1318: Compilation of generated LLVM IR sources is broken
1. Usage a generator to generate LLVM IR
2. Compile the resulting LLVM IR
3. Compilation succeeds, but the binary doesn't contain any output

Example: https://github.com/npmccallum/whitfield/blob/master/meson.build

### PR DESCRIPTION
**Pull request by nirbheek**
Closes #1318

### CODE REVIEW

File: test cases/common/134 generated llvm ir/copy.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env python
```
Comment (ignatenkobrain) [inline test cases/common/134 generated llvm ir/copy.py]: not really sure.... ;)

File: test cases/common/134 generated llvm ir/copy.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+import sys
+import shutil
+
+shutil.copyfile(sys.argv[1], sys.argv[2])
```
Comment (ignatenkobrain) [inline test cases/common/134 generated llvm ir/copy.py]: any reason not to just use some shell script?

File: test cases/common/134 generated llvm ir/main.c [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#include <stdio.h>
+
+unsigned square_unsigned (unsigned a);
+
+int
+main (int argc, char * argv[])
+{
+  unsigned int ret = square_unsigned (2);
+  if (ret != 4) {
+    printf("Got %u instead of 4\n", ret);
```
Comment (ignatenkobrain) [inline test cases/common/134 generated llvm ir/main.c]: what about printing to stderr instead?

File: test cases/common/134 generated llvm ir/copy.py (hunk 1 above)
Comment (nirbheek) [inline test cases/common/134 generated llvm ir/copy.py]: See #1315.

File: test cases/common/134 generated llvm ir/copy.py (hunk 2 above)
Comment (nirbheek) [inline test cases/common/134 generated llvm ir/copy.py]: Windows. (Yes, Clang exists on Windows too.)

File: test cases/common/134 generated llvm ir/main.c (hunk 3 above)
Comment (nirbheek) [inline test cases/common/134 generated llvm ir/main.c]: Doesn't really matter in tests, and this is actually copied from `tests/common/126*/main.c`.

**Comment by jpakkane** [comment]
Looks good. I'm just wondering if we should put all LLVM IR tests in a single directory (that being 126). The number of test projects is starting to get so big that navigating is tricky.

**Comment by nirbheek** [comment]
I didn't put it in 126 because it's already a mess of platform-specific ifs and fors, and because I thought it would be useful to have generation in a separate test like for other features.

We should probably reorganise the tests so that it's easier to figure out what each test does though, and to find the test for a particular feature. Right now the only way is by recursive grepping...

**Comment by jpakkane** [comment]
Maybe we should have a top level test directory "llvm" and move all llvm ir tests there and only run it if `clang` is in path`? Does LLVM have other custom functionality that we might want to add support/tests for?

**Comment by nirbheek** [comment]
There's `llc` support, which I talked with @npmccallum on IRC. See https://github.com/mesonbuild/meson/issues/1317. In that context it definitely makes sense to do that.

**Pull request merged**